apps/server/src/app.ts

```diff
@@ -324,3 +330,46 @@ process.on('uncaughtException', async (error) => {
 process.once('SIGHUP', async () => shutdown(0));
 process.once('SIGINT', async () => shutdown(0));
 process.once('SIGTERM', async () => shutdown(0));
+
+/**
+ * @description tries to open the server with the desired port, and if getting a `EADDRINUSE` will change to an efemeral port
+ * @param {http.Server}server http server object
+ * @param {number}desiredPort the desired port
+ * @returns {number} the resulting port number
+ * @throws any other server errors will result in a throw
+ */
+async function serverTryDesiredPort(server: http.Server, desiredPort: number): Promise<number> {
+  return new Promise((res) => {
+    expressServer.once('error', (e) => {
+      if (isDocker) throw e; // we should only move ports if we are in a desktp environment
+      if (testForPortInUser(e)) {
+        server.listen(0, '0.0.0.0', () => {
+          // @ts-expect-error TODO: find proper documentation for this api
+          const port: number = server.address().port;
+
+          logger.error(
+            LogOrigin.Server,
+            `Failed open the desired port: ${desiredPort} \nMoved to an Ephemeral port: ${port}`,
+            true,
+          );
+
+          res(port);
+        });
+      } else {
+        throw e;
```

Comment:
question: should we reject the promise instead of throwing?

Reply:
there we reach the limit of my knowledge
the goal is to halt the program, how we do that best?

Reply:
the promise constructor gives you two callbacks, reject and resolve (see [docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise))
you are using the resolve one, you could just use the reject instead of throwing

something like (untested)

```ts
async function serverTryDesiredPort(server: http.Server, desiredPort: number): Promise<number> {
  return new Promise((resolve, reject) => {
    expressServer.once('error', (error) => {
      if (isDocker) reject({message: error}); // <-----
      if (testForPortInUser(error)) {
        server.listen(0, '0.0.0.0', () => {
          const port: number = server.address().port;

          logger.error(
            LogOrigin.Server,
            `Failed open the desired port: ${desiredPort} \nMoved to an Ephemeral port: ${port}`,
            true,
          );

          resolve(port);
        });
      } else {
        reject({message: error});  // <-----
      }
    });
    server.listen(desiredPort, '0.0.0.0', () => {
      const port: number = server.address().port;
      resolve(port);
    });
  });
}
```